### PR TITLE
GO-2372 Backlinks & Relations updating mechanism

### PR DESCRIPTION
**Pull request by KirillSto**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### Description

This PR renames system relations Links and Backlinks and makes them unhidden.
Also new logic is introduced that syncs system relations and types presented in spaces with Marketplace space

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Related Tickets & Documents

https://linear.app/anytype/issue/GO-2372/backlinks

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: core/block/object/systemobjectupdate/updater.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+func (u *SystemObjectUpdater) Close(_ context.Context) error {
+	return nil
+}
+
+func (u *SystemObjectUpdater) updateSystemObjects() {
+	marketRels, err := u.store.ListAllRelations(addr.AnytypeMarketplaceWorkspace)
+	if err != nil {
+		log.Errorf("failed to get relations from marketplace space")
```
Comment (AnastasiaShemyakinskaya) [inline core/block/object/systemobjectupdate/updater.go:65]: Why don't we include err in logging? The same question about line 71 and 77

File: core/block/object/systemobjectupdate/updater.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		id := pbtypes.GetString(rec.Details, bundle.RelationKeyId.String())
+		details[id] = rec.Details
+	}
+	return details, nil
+}
+
+func (u *SystemObjectUpdater) updateSystemRelations(spaceId string, marketRels relationutils.Relations) {
+	rels, err := u.store.ListAllRelations(spaceId)
+	if err != nil {
+		log.Errorf("failed to get relations for space %s", spaceId)
```
Comment (AnastasiaShemyakinskaya) [inline core/block/object/systemobjectupdate/updater.go:117]: The same question about logging

File: core/block/object/systemobjectupdate/updater.go (hunk 1 above)
Comment (KirillSto) [reply core/block/object/systemobjectupdate/updater.go]: Thank you! Fixed

File: core/block/object/systemobjectupdate/updater.go (hunk 2 above)
Comment (KirillSto) [reply core/block/object/systemobjectupdate/updater.go]: Added error messages to all logging

**Review by AnastasiaShemyakinskaya — APPROVED** [review]

File: core/block/object/objectcreator/systemobjectreviser.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+	details := make(map[string]*types.Struct, len(records))
+	for _, rec := range records {
+		id := pbtypes.GetString(rec.Details, bundle.RelationKeyId.String())
+		details[id] = rec.Details
+	}
+	return details, nil
+}
+
+func reviseSystemRelation(space space.Space, rel *relationutils.Relation, marketRels relationutils.Relations) {
```
Comment (deff7) [inline core/block/object/objectcreator/systemobjectreviser.go:73]: Maybe just use details structure? realationutils.Relation seems legacy to me

File: core/block/object/objectcreator/systemobjectreviser.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			log.Errorf("failed to update system relation %s in space %s: %v", rel.Key, space.Id(), err)
+		}
+	}
+}
+
+func reviseSystemObjectType(space space.Space, objectType *types.Struct, marketTypes map[string]*types.Struct) {
+	marketType, found := marketTypes[pbtypes.GetString(objectType, bundle.RelationKeySourceObject.String())]
+	rawKey := pbtypes.GetString(objectType, bundle.RelationKeyUniqueKey.String())
+	uk, err := domain.UnmarshalUniqueKey(rawKey)
+	if !found || err != nil || !lo.Contains(bundle.SystemTypes, domain.TypeKey(uk.InternalKey())) ||
```
Comment (deff7) [inline core/block/object/objectcreator/systemobjectreviser.go:95]: maybe split if condition and move each check closely to places where corresponding variables appear?

```
marketType, found := marketTypes[pbtypes.GetString(objectType, bundle.RelationKeySourceObject.String())]
if !found {
  return
}
...
uk, err := domain.UnmarshalUniqueKey(rawKey)
if err != nil {
  return
}
...etc
```

File: core/block/object/objectcreator/systemobjectreviser.go (hunk 3 above)
Comment (KirillSto) [reply core/block/object/objectcreator/systemobjectreviser.go]: Done!

File: core/block/object/objectcreator/systemobjectreviser.go (hunk 4 above)
Comment (KirillSto) [reply core/block/object/objectcreator/systemobjectreviser.go]: I have moved out checks on belonging to system objects to separate func, so it is used both for types and relations

**Review by deff7 — APPROVED** [review]

**Pull request merged**